renderdoc/driver/vulkan/vk_pixelhistory.cpp

```diff
@@ -0,0 +1,607 @@
+                             VkFormat format, uint32_t sampleMask, VkQueryPool occlusionPool,
+                             VkImageView colorImageView, VkImageView stencilImageView,
+                             VkImage colorImage, VkImage stencilImage, VkBuffer dstBuffer,
+                             const std::vector<EventUsage> &events)
+      : m_pDriver(vk),
+        m_X(x),
+        m_Y(y),
+        m_Image(image),
+        m_Format(format),
+        m_DstBuffer(dstBuffer),
+        m_EventIndex(0),
+        m_SampleMask(sampleMask),
+        m_OcclusionPool(occlusionPool),
+        m_ColorImageView(colorImageView),
+        m_StencilImageView(stencilImageView),
+        m_ColorImage(colorImage),
+        m_StencilImage(stencilImage),
+        m_PrevState(vk, NULL)
+  {
+    m_pDriver->SetDrawcallCB(this);
+    for(size_t i = 0; i < events.size(); i++)
+      m_Events.insert(std::make_pair(events[i].eventId, events[i]));
+  }
+  ~VulkanPixelHistoryCallback() { m_pDriver->SetDrawcallCB(NULL); }
+  void CopyPixel(uint32_t eid, VkCommandBuffer cmd, uint32_t offset)
+  {
+    CopyPixelParams colourCopyParams = {};
+    colourCopyParams.multisampled = false;    // TODO: multisampled
+    colourCopyParams.srcImage = m_Image;
+    colourCopyParams.srcImageLayout =
+        VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL;    // TODO: image layout
+    colourCopyParams.srcImageFormat = m_Format;
+    colourCopyParams.imageOffset = VkOffset3D{int32_t(m_X), int32_t(m_Y), 0};
+    colourCopyParams.dstBuffer = m_DstBuffer;
+
+    m_pDriver->GetDebugManager()->PixelHistoryCopyPixel(cmd, colourCopyParams, offset);
+
+    const DrawcallDescription *draw = m_pDriver->GetDrawcall(eid);
+    if(draw && draw->depthOut != ResourceId())
+    {
+      // The draw call had a depth image attachment.
+      ResourceId depthImage = m_pDriver->GetResourceManager()->GetLiveID(draw->depthOut);
+      VulkanCreationInfo::Image imginfo = m_pDriver->GetDebugManager()->GetImageInfo(depthImage);
+      CopyPixelParams depthCopyParams = colourCopyParams;
+      depthCopyParams.depthcopy = true;
+      depthCopyParams.srcImage =
+          (VkImage)m_pDriver->GetResourceManager()->GetCurrentResource(depthImage);
+      depthCopyParams.srcImageLayout = VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL;
+      depthCopyParams.srcImageFormat = imginfo.format;
+      m_pDriver->GetDebugManager()->PixelHistoryCopyPixel(cmd, depthCopyParams, offset + 16);
+    }
+  }
+
+  void PreDraw(uint32_t eid, VkCommandBuffer cmd)
+  {
+    auto it = m_Events.find(eid);
+    if(it == m_Events.end())
+      return;
+    EventUsage event = it->second;
+    m_PrevState = m_pDriver->GetRenderState();
+
+    m_pDriver->GetRenderState().EndRenderPass(cmd);
+    // Get pre-modification values
+    uint32_t storeOffset = m_EventIndex * sizeof(EventInfo);
+    CopyPixel(eid, cmd, storeOffset);
+
+    // TODO: add all logic for draw calls.
+
+    m_pDriver->GetRenderState() = m_PrevState;
```

Comment:
Is this push/pop to `m_PrevState` necessary? I think any code we run in `CopyPixel` shouldn't change the stored RenderState, and the begin below will re-set everything necessary.

Reply:
Removed, initially where I had a TODO I had code that was modifying the state, so this was leftover from then